pageserver/src/tenant/timeline.rs

```diff
@@ -3642,12 +3589,54 @@ impl Timeline {
             }
             prev = Some(next_key.next());
         }
+        stats.read_lock_held_compute_holes_micros =
+            stats.read_lock_held_prerequisites_micros.till_now();
         drop(layers);
-        stats.second_read_lock_held_micros = stats.second_read_lock_acquisition_micros.till_now();
+        stats.read_lock_drop_micros = stats.read_lock_held_compute_holes_micros.till_now();
         let mut holes = heap.into_vec();
         holes.sort_unstable_by_key(|hole| hole.key_range.start);
         let mut next_hole = 0; // index of next hole in holes vector
-        stats.sort_holes_micros = stats.second_read_lock_held_micros.till_now();
+
+        // This iterator walks through all key-value pairs from all the layers
+        // we're compacting, in key, LSN order.
+        let all_values_iter = itertools::process_results(
+            deltas_to_compact.iter().map(|l| l.iter(ctx)),
+            |iter_iter| {
+                iter_iter.kmerge_by(|a, b| {
+                    if let Ok((a_key, a_lsn, _)) = a {
+                        if let Ok((b_key, b_lsn, _)) = b {
+                            match a_key.cmp(b_key) {
+                                Ordering::Less => true,
+                                Ordering::Equal => a_lsn <= b_lsn,
```

Comment:
This is not changed by this PR but I'm wondering why it isn't `a_lsn < b_lsn`? Not sure if using `=` here will cause any unexpected behavior. (Probably it is not possible to have two keys of the same lsn in the current setup unless something goes wrong when deleting a layer?)

Reply:
My undestanding is that a single ingested WAL record results in one or more `Timeline::put`s for different keys at the WAL record's LSN.

The next WAL record will have a different LSN.

Hence it doesn't matter. But let's confirm with @knizhnik ; however, I will merge this PR regardless as it's a pre-existing condition.

Reply:
Yes, right now key is assumed to be unique.
In theory it is possible to that ingest of one WAL record cause storing value for some key multiple time.
But I didn't see wal redo handlers which now can behave in this way.
Also if Key is  not unique, then current sorting data by key will not be correct in any case, doesn't matter whether `<` ot `<=` comparison is used: we need stable sort.